plugins/hls-cabal-plugin/test/CabalAdd.hs

```diff
@@ -1,32 +1,102 @@
+                                                                testGroup,
+                                                                waitForDiagnosticsFrom,
+                                                                (@?=))
 import           Utils
 
-cabalAddTests :: TestTree
-cabalAddTests =
+cabalAddModuleTests :: TestTree
+cabalAddModuleTests =
+    testGroup
+        "Add Module"
+        [ runHaskellTestCaseSession "Code Actions - Can add module to benchmark" ("cabal-add-module" </> "library") $ do
+            let compName = CBenchName "test1"
+            pd <- generateAddDependencyTestSession "test.cabal" "Main.hs" compName
+            checkModuleAddedTo pd "Main" $ compName
+        , runHaskellTestCaseSession "Code Actions - Can add module to executable" ("cabal-add-module" </> "library") $ do
+            let compName = CExeName "test"
+            pd <- generateAddDependencyTestSession "test.cabal" "Main.hs" compName
+            checkModuleAddedTo pd "Main" $ compName
+        , runHaskellTestCaseSession "Code Actions - Can add module to test-suite" ("cabal-add-module" </> "library") $ do
+            let compName = CTestName "test2"
+            pd <- generateAddDependencyTestSession "test.cabal" "Main.hs" compName
+            checkModuleAddedTo pd "Main" $ compName
+        , runHaskellTestCaseSession "Code Actions - Can add module to library" ("cabal-add-module" </> "library") $ do
+            let compName = CLibName $ LSubLibName "test3"
+            pd <- generateAddDependencyTestSession "test.cabal" "Main.hs" compName
+            checkModuleAddedTo pd "Main" $ compName
+        , runHaskellTestCaseSession "Code Actions - Can add module to main library" ("cabal-add-module" </> "library") $ do
+            let compName = CLibName LMainLibName
+            pd <- generateAddDependencyTestSession "test.cabal" "Main.hs" compName
+            checkModuleAddedTo pd "Main" $ compName
+        ]
+  where
+    generateAddDependencyTestSession :: FilePath -> FilePath -> ComponentName -> Session PackageDescription
+    generateAddDependencyTestSession cabalFile haskellFile compName = do
+        haskellDoc <- openDoc haskellFile "haskell"
+        cabalDoc <- openDoc cabalFile "cabal"
+        _ <- waitForDiagnosticsFrom haskellDoc
+        cas <- Maybe.mapMaybe (^? _R) <$> getAllCodeActions haskellDoc
+        let selectedCas = filter (\ca -> (T.pack $ "Add to " <> (Pretty.prettyShow compName)) `T.isPrefixOf` (ca ^. L.title)) cas
+        mapM_ executeCodeAction $ selectedCas
+        _ <- skipManyTill anyMessage $ getDocumentEdit cabalDoc -- Wait for the changes in cabal file
+        contents <- documentContents cabalDoc
+        case parseCabalFileContents $ T.encodeUtf8 contents of
+            (_, Right gpd) -> pure $ flattenPackageDescription gpd
+            _ -> liftIO $ assertFailure "could not parse cabal file to gpd"
+
+    checkModuleAddedTo :: PackageDescription -> String -> ComponentName -> Session ()
```

Comment:
maybe document that we do not care whether a file is added to exposed or other modules